gobblin-modules/gobblin-eventhub/src/main/java/org/apache/gobblin/eventhub/source/EventhubSource.java

```diff
@@ -0,0 +1,152 @@
+import org.apache.gobblin.source.workunit.WorkUnit;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.*;
+
+
+/**
+ * Created by Jan on 23-Jan-19.
+ */
+public class EventhubSource extends AbstractSource<Void, EventData> {
+
+    public static final Logger LOG = LoggerFactory.getLogger(EventhubSource.class);
+
+    public static final String WORK_UNIT_STATE_VERSION_KEY = "source.querybased.workUnitState.version";
+    public static final Integer CURRENT_WORK_UNIT_STATE_VERSION = 1;
+
+    @Override
+    public List<WorkUnit> getWorkunits(SourceState state) {
+        List<WorkUnit> workUnits = new ArrayList<>();
+
+        Long previousWatermark = getPreviousWatermark(state);
+        if (previousWatermark == null) {
+            previousWatermark = ConfigurationKeys.DEFAULT_WATERMARK_VALUE;
+        }
+
+        // for now, enforce single work unit to simplify watermark handling:
+        if (state.getPropAsInt(ConfigurationKeys.SOURCE_MAX_NUMBER_OF_PARTITIONS, 1) != 1) {
+            LOG.error("Unsupported number of partitions, resetting to 1.");
+        }
+        state.setProp(ConfigurationKeys.SOURCE_MAX_NUMBER_OF_PARTITIONS, 1);
+        workUnits.addAll(generateWorkUnits(state, previousWatermark));
+
+        LOG.info("Total number of workunits for the current run: " + workUnits.size());
+        List<WorkUnit> previousWorkUnits = this.getPreviousWorkUnitsForRetry(state);
+        LOG.info("Total number of incomplete tasks from the previous run: " + previousWorkUnits.size());
+        workUnits.addAll(previousWorkUnits);
+
+        int numOfMultiWorkunits =
+                state.getPropAsInt(ConfigurationKeys.MR_JOB_MAX_MAPPERS_KEY, ConfigurationKeys.DEFAULT_MR_JOB_MAX_MAPPERS);
+
+        return pack(workUnits, numOfMultiWorkunits);
+    }
+
+    private static List<WorkUnit> pack(List<WorkUnit> workUnits, int numOfMultiWorkunits) {
+        Preconditions.checkArgument(numOfMultiWorkunits > 0);
+
+        if (workUnits.size() <= numOfMultiWorkunits) {
+            return workUnits;
+        }
+        List<WorkUnit> result = Lists.newArrayListWithCapacity(numOfMultiWorkunits);
+        for (int i = 0; i < numOfMultiWorkunits; i++) {
+            result.add(MultiWorkUnit.createEmpty());
+        }
+        for (int i = 0; i < workUnits.size(); i++) {
+            ((MultiWorkUnit) result.get(i % numOfMultiWorkunits)).addWorkUnit(workUnits.get(i));
+        }
+        return result;
+    }
+
+    protected List<WorkUnit> generateWorkUnits(SourceState state, long previousWatermark) {
+        List<WorkUnit> workUnits = new ArrayList<>();
+
+        state.setProp(ConfigurationKeys.EXTRACT_DELTA_FIELDS_KEY, "offset");
+        state.setProp(ConfigurationKeys.SOURCE_QUERYBASED_WATERMARK_TYPE, "simple");
+        state.setProp(ConfigurationKeys.SOURCE_QUERYBASED_EXTRACT_TYPE, "snapshot");
+        List<Partition> partitions = new Partitioner(state).getPartitionList(previousWatermark);
```

Comment:
Like I mentioned, there shouldn't be a global previous watermark. Right now if you have one partition, it will cut one partition range into multiple pieces, which doesn't have any ordering guarantees. This is not what we want.